unixfs/hamt/hamt.go

```diff
@@ -0,0 +1,466 @@
+		if err != nil {
+			return err
+		}
+
+		lnk.Name = sv.key
+
+		return f(lnk)
+	})
+}
+
+func (ds *HamtShard) walkTrie(cb func(*shardValue) error) error {
+	for i := 0; i < ds.tableSize; i++ {
+		if ds.bitfield.Bit(i) == 0 {
+			continue
+		}
+
+		idx := ds.indexForBitPos(i)
+		// NOTE: an optimized version could simply iterate over each
+		//       element in the 'children' array.
+		c, err := ds.getChild(context.TODO(), idx)
+		if err != nil {
+			return err
+		}
+
+		switch c := c.(type) {
+		case *shardValue:
+			err := cb(c)
+			if err != nil {
+				return err
+			}
+
+		case *HamtShard:
+			err := c.walkTrie(cb)
+			if err != nil {
+				return err
+			}
+		default:
+			return fmt.Errorf("unexpected child type: %#v", c)
+		}
+	}
+	return nil
+}
+
+func (ds *HamtShard) modifyValue(ctx context.Context, hv *hashBits, key string, val node.Node) error {
+	idx := hv.Next(ds.tableSizeLg2)
+
+	if ds.bitfield.Bit(idx) != 1 {
+		return ds.insertChild(idx, key, val)
+	}
+
+	cindex := ds.indexForBitPos(idx)
+
+	child, err := ds.getChild(ctx, cindex)
+	if err != nil {
+		return err
+	}
+
+	switch child := child.(type) {
+	case *HamtShard:
+		err := child.modifyValue(ctx, hv, key, val)
+		if err != nil {
+			return err
+		}
+
+		if val == nil {
+			switch len(child.children) {
+			case 0:
+				// empty sub-shard, prune it
+				// Note: this shouldnt normally ever happen
+				//       in the event of another implementation creates flawed
+				//       structures, this will help to normalize them.
+				ds.bitfield.SetBit(ds.bitfield, idx, 0)
+				return ds.rmChild(cindex)
+			case 1:
+				nchild, ok := child.children[0].(*shardValue)
+				if ok {
+					// sub-shard with a single value element, collapse it
+					ds.setChild(cindex, nchild)
+				}
+				return nil
+			}
+		}
+
+		return nil
+	case *shardValue:
+		switch {
+		case val == nil: // passing a nil value signifies a 'delete'
+			ds.bitfield.SetBit(ds.bitfield, idx, 0)
+			return ds.rmChild(cindex)
+
+		case child.key == key: // value modification
+			child.val = val
+			return nil
+
+		default: // replace value with another shard, one level deeper
+			ns := NewHamtShard(ds.dserv, ds.tableSize)
+			chhv := &hashBits{
+				b:        hash([]byte(child.key)),
+				consumed: hv.consumed,
+			}
+
+			err := ns.modifyValue(ctx, hv, key, val)
+			if err != nil {
+				return err
+			}
+
+			err = ns.modifyValue(ctx, chhv, child.key, child.val)
+			if err != nil {
+				return err
+			}
+
+			ds.setChild(cindex, ns)
+			return nil
+		}
+	default:
+		return fmt.Errorf("unexpected type for child: %#v", child)
+	}
+}
+
+func (ds *HamtShard) indexForBitPos(bp int) int {
+	// TODO: an optimization could reuse the same 'mask' here and change the size
+	//       as needed. This isnt yet done as the bitset package doesnt make it easy
+	//       to do.
+	mask := new(big.Int).Sub(new(big.Int).Exp(big.NewInt(2), big.NewInt(int64(bp)), nil), big.NewInt(1))
+	mask.And(mask, ds.bitfield)
+
+	return popCount(mask)
```

Comment:
I'm not entirely clear on how this function works/what exactly it's purpose is. And in turn I'm having a hard time verifying that the bit operations you do, do what they are supposed to be doing. Would be great to have a comment about what the function does and an explenation why the bit ops do what they do.